### PR TITLE
EF6 to EF Core porting guide

### PR DESCRIPTION
**Pull request by JeremyLikness**
A comprehensive guide to porting from EF6 to EF Core. This will address #1180.

### CODE REVIEW

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/index.md:18]: I'm never sure these days how to unambiguously refer to what used to be .NET Core. Since .NET 5 and .NET 6 are not ".NET Core", then it might be confusing to just say .NET Core here. Maybe explicitly call this out in some way? (Although everything I can think of sounds clumsy! Why did we rename again?)

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/index.md:22]: We should update and link to the things that EF Core supports that EF6 does not. (And vice-versa, but calling this out here because many people seem to be unaware of the many fundamental things that are build into EF Core but will never be supported in EF6; example relationships using alternate keys.)

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/index.md:25]: Maybe add using Dependency Injection here?

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/index.md:30]: Not sure if I would say they are "fundamentally different". More that the implementation has been significantly improved, even though they are fundamentally the same thing.

Comment (ajcvickers) [reply entity-framework/efcore-and-ef6/porting/index.md:30]: This is good, but we should plan on providing more detailed instructions here. (Probably already done below, or planned.)

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/index.md:44]: I love using "database is the source of truth" and "code is the source of truth" here. The hybrid mapping approach is maybe a bit harder to grok. Maybe save it for a more advanced section later? (Just ideas.)

File: entity-framework/efcore-and-ef6/porting/port-behavior.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-behavior.md:21]: Would be good to link to other docs where they have more information--change tracking in this case.

**Comment by JeremyLikness** [comment]
@dotnet/efteam this is in draft status ready for your review.

I'll tag some other individuals after this first round.

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/index.md:3]: 👀

Comment (JeremyLikness) [reply entity-framework/efcore-and-ef6/porting/index.md]: "Nothing to see here"

File: entity-framework/efcore-and-ef6/porting/port-code.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-code.md:26]: Update the date for this page

File: entity-framework/efcore-and-ef6/porting/port-database.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-database.md:12]:
```suggestion
If you're using the database as the source of truth, the upgrade will mostly involve addressing any changes to the shape of entities generated. The steps to migrate include:
```

File: entity-framework/efcore-and-ef6/porting/port-detailed-cases.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:27]: Consider rephrasing the last sentence, also it's possible to get a model builder already initialized with the conventions

File: entity-framework/efcore-and-ef6/porting/port-detailed-cases.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:47]: Consider breaking out spatial to its own section

File: entity-framework/efcore-and-ef6/porting/port-detailed-cases.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:57]: Explicitly call out that assemblies are not scanned for derived types

File: entity-framework/efcore-and-ef6/porting/port-detailed-cases.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:61]: The example is not relevant.  `Map` is usually used to configure inheritance mapping, so as an example `HasDiscriminator` can be used to configure TPH

File: entity-framework/efcore-and-ef6/porting/port-detailed-cases.md
Comment (AndriySvyryd) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:65]: Link to relevant documentation pages here and other places

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ErikEJ) [inline entity-framework/efcore-and-ef6/porting/index.md:22]: "will only be added to"

File: entity-framework/efcore-and-ef6/porting/port-database.md
Comment (ErikEJ) [inline entity-framework/efcore-and-ef6/porting/port-database.md:38]: There are also third party tools, that offer additional ways of customizing the generated code. (with link to tools page?)

**Comment by JeremyLikness** [comment]
Adding @ryanbrandenburg and @jinujoseph for review and feedback.

**Comment by JeremyLikness** [comment]
Hi @dotnet/efteam and @ErikEJ ... I made a pass at updates, waiting for an approval or another round of edits. Goal is to get this out early next week. Thanks!

File: entity-framework/efcore-and-ef6/porting/port-database.md
Comment (ryanbrandenburg) [inline entity-framework/efcore-and-ef6/porting/port-database.md:39]: Surely this is not a complete list (which is guaranteed to stay that way)? Maybe more like:
`For a list of featured community tools and extensions, ...`

Comment (JeremyLikness) [reply entity-framework/efcore-and-ef6/porting/port-database.md]: Good point! Fixed.

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/index.md:65]: "Move from `Database.Log` to `Microsoft.Extensions.Logging APIs`". Moving to simple logging may be easier. Probably should suggest both. For example, "Move from `Database.Log` to `Microsoft.Extensions.Logging` APIs or `DbContextOptionsBuilder.LogTo`"

(Also, move code fence tick before APIs.)

File: entity-framework/efcore-and-ef6/porting/port-behavior.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-behavior.md:17]: Couple of comments:
- This change of behavior only applies to the `Attach` and `Update` method groups. `Add` always puts entities in the `Added` state, even if the key is set.
- `Attach` methods put entities with keys set into the `Unchanged` state. This facilitates "insert it if new, otherwise leave it alone." `Update` methods put entities with keys set into the `Modified` state. This facilitates "insert it if new, otherwise update it."

The general philosophy here is that `Update` is a very simple way to handle inserts and updates of disconnected entities. It ensures any new entities are inserted, and any existing entities are updated.

At the same time, `Add` still provides an easy way force entities to be inserted. `Add` is mostly useful only when not using store-generated keys, such that EF doesn't know if the entity is new or not.

File: entity-framework/efcore-and-ef6/porting/port-behavior.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-behavior.md:31]: I think, "is in the applications `App/Web.config` file" should be "is in the application's `App/Web.config` file."

Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-behavior.md:31]: Link to "Managing database schemas" at the bottom of this section?

File: entity-framework/efcore-and-ef6/porting/port-database.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-database.md:37]: "As of EF Core 6, it is possible to customize..." It is currently **not** possible.

File: entity-framework/efcore-and-ef6/porting/port-database.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-database.md:17]: I think we should reference Erik's Power Tools here.

File: entity-framework/efcore-and-ef6/porting/port-detailed-cases.md
Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:71]: "For spatial support, EF Core recommends using a third-party library such as NetTopologySuite." I would change this to indicate that EF Core integrates with NTS. Using a different third-party library won't work unless this same integration work is done, which is non-trivial. Something like, "For spatial support, EF Core integrates with the third-party community library NetTopologySuite."

Comment (ajcvickers) [inline entity-framework/efcore-and-ef6/porting/port-detailed-cases.md:71]: I would explicitly mention in the mapping section that inherited types of a mapped entity type are not discovered automatically by EF Core. This has tripped up many people in the past.

**Review by ajcvickers — APPROVED** [review]
Love it!

File: entity-framework/efcore-and-ef6/porting/index.md
Comment (JeremyLikness) [reply entity-framework/efcore-and-ef6/porting/index.md:65]: great point!

**Pull request merged**
